clang/lib/Index/IndexRecordHasher.cpp

```diff
@@ -184,6 +194,26 @@ void IndexRecordHasher::hashMacro(const IdentifierInfo *name,
 void IndexRecordHasher::hashDecl(const Decl *D) {
   assert(D->isCanonicalDecl());
 
+  auto emplaceResult = HashedDecls.try_emplace(D, HashedDecls.size() + 1);
+  bool inserted = emplaceResult.second;
+  if (!inserted) {
+    // If we have already serialized this declaration, just add the
+    // declaration's hash to the hash builder. This is significantly
+    // cheaper than visiting the declaration again.
+    HashBuilder.add(emplaceResult.first->second);
+    return;
+  }
+
+  // If we haven't serialized the declaration yet,`try_emplace` will insert the
+  // new unique ID into `HashedDecls`. We just need to hash the declaration
+  // once.
+  //
+  // Add a `0` to the hash value to indicate that we are serializing a new
+  // declaration. The declaration IDs serialized above is always greater than
+  // zero, so there is no ambiguity between newly serialized decls and the
+  // serialization of declaration IDs.
+  HashBuilder.add(0);
```

Comment:
Doesn't really matter, but is this actually needed 🤔?

Reply:
That seems exceedingly unlikely to me as we end up adding the type with a `#` in first. But 🤷 Happy to keep if you really want it